linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -1971,6 +1972,26 @@ private void closeInner() {
       final Function2<TSource, TInner, TResult> resultSelector,
       boolean generateNullsOnLeft,
       boolean generateNullsOnRight) {
+    return mergeJoin(outer, inner, outerKeySelector, innerKeySelector, null,
+        resultSelector, generateNullsOnLeft, generateNullsOnRight);
+  }
+
+  /**
+   * Joins two inputs that are sorted on the key, with an extra predicate for non equi-join
+   * conditions (in case of equi-join, it will be null).
+   * Inputs must sorted in ascending order, nulls last.
+   * NOTE: The current API is experimental and subject to change without notice.
+   */
+  @Experimental
+  public static <TSource, TInner, TKey extends Comparable<TKey>, TResult> Enumerable<TResult>
+      mergeJoin(final Enumerable<TSource> outer,
+      final Enumerable<TInner> inner,
+      final Function1<TSource, TKey> outerKeySelector,
+      final Function1<TInner, TKey> innerKeySelector,
+      final Predicate2<TSource, TInner> extraPredicate,
```

Comment:
Please add the relevant documentation that clarifies why this `extraPredicate` is needed and how it is different from a predicate on top of join results.

Reply:
You're right, I'll document that we use a predicate rather that a filter on top of join results in order to support other join types in the (near) future.

Reply:
documentation updated

Reply:
Thanks. However, can you please add `@param extraPredicate` that would clarify the meaning of `extraPredicate` in a couple of words?

The documentation does not need to clarfy all the implementation details, however, the naming of `Predicate2<TSource, TInner> extraPredicate` provides absolutely no clue one meaning of the predicate.

Well, I do see it is a predicate. I can judge that by class name. As a user of that API I want to know what is expected to be passed there. I want to know if `nulls` are allowed or not and so on.

Reply:
javadoc updated: added `@param extraPredicate`